### PR TITLE
flags: Clone to avoid modifying os.Args

### PR DESCRIPTION
**Pull request by mqasimsarfraz**
If we try to shrink args (to drop '--sort ...') e.g from `top file --sort -writes -m 10 --timeout 10 --interval 10` we end up reusing the underlying array and hence the original os.Args end up getting modified resulting in `top file -m 10 --timeout 10 --interval 10 --interval 10`. The commit clones the slice to ensure os.Args stay untouched and evaluated correctly later on.

### before the PR

```
→ ./kubectl-gadget  snapshot process --sort pid  --all-namespaces
K8S.NODE                                   K8S.NAMESPACE                             K8S.POD                                   K8S.CONTAINER                             COMM             PID                  UID        GID       
minikube-docker                            gadget                                    gadget-76f66                              gadget                                    gadgettracerman  3591435              0          0         
minikube-docker                            kube-system                               storage-provisioner                       storage-provisioner                       storage-provisi  3269857              0          0         
minikube-docker                            kube-system                               coredns-6d4b75cb6d-9dsjr                  coredns                                   coredns          3269733              0          0         
minikube-docker                            kube-system                               kube-proxy-swnds                          kube-proxy                                kube-proxy       3269478              0          0         
minikube-docker                            kube-system                               kube-apiserver-minikube-docker            kube-apiserver                            kube-apiserver   3268122              0          0         
minikube-docker                            kube-system                               kube-controller-manager-minikube-docker   kube-controller-manager                   kube-controller  3268080              0          0         
minikube-docker                            kube-system                               etcd-minikube-docker                      etcd                                      etcd             3268070              0          0         
minikube-docker                            kube-system                               kube-scheduler-minikube-docker            kube-scheduler                            kube-scheduler   3268062              0          0   
```

### After the PR

```
→ ./kubectl-gadget  snapshot process --sort pid  --all-namespaces
K8S.NODE                                   K8S.NAMESPACE                             K8S.POD                                   K8S.CONTAINER                             COMM             PID                  UID        GID       
minikube-docker                            kube-system                               kube-scheduler-minikube-docker            kube-scheduler                            kube-scheduler   3268062              0          0         
minikube-docker                            kube-system                               etcd-minikube-docker                      etcd                                      etcd             3268070              0          0         
minikube-docker                            kube-system                               kube-controller-manager-minikube-docker   kube-controller-manager                   kube-controller  3268080              0          0         
minikube-docker                            kube-system                               kube-apiserver-minikube-docker            kube-apiserver                            kube-apiserver   3268122              0          0         
minikube-docker                            kube-system                               kube-proxy-swnds                          kube-proxy                                kube-proxy       3269478              0          0         
minikube-docker                            kube-system                               coredns-6d4b75cb6d-9dsjr                  coredns                                   coredns          3269733              0          0         
minikube-docker                            kube-system                               storage-provisioner                       storage-provisioner                       storage-provisi  3269857              0          0         
minikube-docker                            gadget                                    gadget-76f66                              gadget                                    gadgettracerman  3591435              0          0     
```
**Note:**  You need `sort` before `--all-namespace` to ensure while shrinking the args we modify the order of items in [`removeSplitSortArgs`](https://github.com/inspektor-gadget/inspektor-gadget/blob/main/cmd/common/utils/flags.go#L163) to reproduce the issue.

### CODE REVIEW

File: .github/workflows/inspektor-gadget.yml
Comment (blanquicet) [inline .github/workflows/inspektor-gadget.yml:5]: Why is this upgrade necessary for this PR?

File: cmd/common/utils/flags.go
Comment (blanquicet) [inline cmd/common/utils/flags.go:192]: Good catch!

This is another solution but I think your proposal is cleaner:

```go
// removeSplitSortArgs removes the --sort flag with its arg, if it isn't in the
// merged form of --sort=foo
func removeSplitSortArgs(args []string) []string {
	for i := 0; i < len(args); i++ {
		if args[i] == "--sort" {
			// Remove also the next element as it is the arg of --sort
			ret := make([]string, 0, len(args)-2)
			ret = append(ret, args[:i]...)
			return append(ret, args[i+2:]...)
		}
	}
	return args
}

// removeHelpArg removes the --help flag
func removeHelpArg(args []string) []string {
	for i := 0; i < len(args); i++ {
		if args[i] == "--help" || args[i] == "-h" {
			ret := make([]string, 0, len(args)-1)
			ret = append(ret, args[:i]...)
			return append(ret, args[i+1:]...)
		}
	}
	return args
}
```

File: .github/workflows/inspektor-gadget.yml
Comment (mqasimsarfraz) [reply .github/workflows/inspektor-gadget.yml:5]: To have `slices.Clone()`, I mean looking at what `slices.Clone()` we can also do (using [append](https://stackoverflow.com/questions/71316774/is-it-equivalent-use-copy-or-append-slice-type-in-golang)):

```args := append([]string{}, os.Args[1:]...) // clone to avoid modifying the original os.Args```

But should be fine to go to `1.21`. What do you say?

File: cmd/common/utils/flags.go
Comment (mqasimsarfraz) [reply cmd/common/utils/flags.go:192]: > This is another solution but I think your proposal is cleaner

Yeah, Also cloning one level above ensure it is done for all following function.

File: .github/workflows/inspektor-gadget.yml
Comment (blanquicet) [reply .github/workflows/inspektor-gadget.yml:5]: IMO, I would avoid upgrading it unless it is strictly necessary or it provides a big advantage. And, I consider this doesn't seem the case.

However, let's see what other maintainers think.

Comment (flyth) [reply .github/workflows/inspektor-gadget.yml:5]: I think 1.21 with its `maps` and `slices` packages and the new `min()` and `max()` funcs alone are worth the upgrade as we use those things constantly. And this is timed just after a release, so there's plenty of time ahead of us to make sure it's running well. So my opinion is: go for it.

The release notes state that Go 1.21 no longer runs on Win 7 & 8 and older MacOS versions, but I don't know if the binaries it produces are affected as well - we should check for that and maybe add that to the release notes.

Comment (mqasimsarfraz) [reply .github/workflows/inspektor-gadget.yml:5]: > (But maybe this should get its own PR prior to this one)

Yeah, maybe it is a better idea to open a separate PR for golang 1.21 and gather more use-cases.

> `args := append([]string{}, os.Args[1:]...) // clone to avoid modifying the original os.Args`

If the [above](https://github.com/inspektor-gadget/inspektor-gadget/pull/2195/files#r1381893852) approach looks good, I can unblock this fix and we can continue with `golang 1.21` separately?

Comment (mauriciovasquezbernal) [reply .github/workflows/inspektor-gadget.yml:5]: I'll suggest not to tie these two things together. Unblock this PR by using that approach and add a comment saying it should be changed once we use go1.21, then in another PR we can discuss the go1.21 implications.

**Comment by mauriciovasquezbernal** [comment]
Perhaps you can also include the commit from https://github.com/inspektor-gadget/inspektor-gadget/pull/2192, so we get all the fixes in the same PR?

File: .github/workflows/inspektor-gadget.yml
Comment (mqasimsarfraz) [reply .github/workflows/inspektor-gadget.yml]: Makes sense. I would open a separate PR for bumping `golang` version. For now I switched to different approach with a comment.

File: cmd/common/utils/flags.go
Comment (mqasimsarfraz) [reply cmd/common/utils/flags.go]: Yep, that should also do the trick. I read [this](https://github.com/go101/go101/wiki/There-is-not-a-perfect-way-to-clone-slices-in-Go) to confuse myself more :D and then decided to go for whatever [slices](https://cs.opensource.google/go/x/exp/+/7918f672:slices/slices.go;l=336) will give us in future.

**Review by flyth — APPROVED** [review]
LGTM, thanks!

**Comment by mqasimsarfraz** [comment]
Thanks for reviews!

File: .github/workflows/inspektor-gadget.yml
Comment (mqasimsarfraz) [reply .github/workflows/inspektor-gadget.yml]: I opened https://github.com/inspektor-gadget/inspektor-gadget/pull/2206 to update golang version.

**Pull request merged**
